2-Draft-Accepted/RFCxxxx-PowerShellGet-3.0.md

```diff
@@ -0,0 +1,242 @@
+---
+RFC: RFCxxxx
+Author: Steve Lee
+Status: Draft
+SupercededBy: N/A
+Version: 0.1
+Area: PowerShellGet
+Comments Due: 7/30
+Plan to implement: Yes
+---
+
+# PowerShellGet 3.0 Module
+
+PowerShellGet has enabled an active community to publish and install PowerShell
+Modules.
+However, it has some architectural and design limitations that prevent it from
+moving forward more quickly.
+This RFC proposes some significant changes to address this.
+
+> [!NOTE]
+> There are no intentions to break existing PSGet v2 users so existing
+> users can stay on PSGet v2 and that will continue to work.
+
+## Motivation
+
+    As a PowerShell user,
+    I can discover how to install missing cmdlets,
+    so that I can be more productive.
+
+    As a PowerShellGet contributor,
+    I can more easily contribute code,
+    so that PowerShellGet evolves more quickly.
+
+## User Experience
+
+```powershell
+PS> Get-Satisfaction
+Get-Satisfaction : The term 'Get-Satisfaction' is not recognized as the name of a cmdlet, function, script file, or operable program.
+Check the spelling of the name, or if a path was included, verify that the path is correct and try again.
+At line:1 char:1
++ Get-Satisfaction
++ ~~~
++ CategoryInfo          : ObjectNotFound: (Get-Satisfaction:String) [], CommandNotFoundException
++ FullyQualifiedErrorId : CommandNotFoundException
+
+Suggestion: You can obtain `Get-Satisfaction` by running `Install-PSResource HappyDays`.
+
+PS> Install-PSResource HappyDays
+```
+
+## Specification
+
+### Rewrite of PowerShellGet
+
+PowerShellGet is currently written as PowerShell script with a dependency on PackageManagement.
+Proposal is to write PSResource in C# to reduce complexity and make easier to maintain.
+In addition, remove dependency on PackageManagement completely as well as dependency on
+nuget.exe.
+This module will only use REST APIs to get and publish nupkgs.
+This module would be shipped in PowerShell 7.
+
+### Side-by-side with PowerShellGet
+
+To ease transition due to the enormity of the breaking change, this module will
+have a different names for the cmdlets to allow for side by side use
+with the existing PowerShellGet module.
+Since the current PowerShellGet 2.x version is a script module and this new one
+is a C# based module, they can coexist side-by-side.
```

Comment:
It would be good to state explicitly whether PowerShellGet 2.x and/or PackageManagement would be included in the PS7.0 distribution or not

Reply:
It depends on if PSGet v3 is done in time for PS7.  Desire is for v3 to be in 7.

Reply:
Has PowerShell ever removed an in-box module outside the technical limitations of the 5.1 to 6.0 release? If so, does that constitute a breaking change since the commands are no longer present by default?

Reply:
Kind of? Some of the windows-specific modules didn't come in the PS Core box, right?

Reply:
- `Microsoft.PowerShell.LocalAccounts`
- `Microsoft.PowerShell.ODataUtils`
- `Microsoft.PowerShell.Operation.Validation`

are PowerShell-namedspaced but not shipped in Core. Of those, the only one I know anything about is `LocalAccounts` and that's Windows-specific, shipped as part of Windows, and is Core-compatible (of which I think make sense -- it's just that the module should be called something like `LocalAccounts` ideally).

There are also a bunch of Windows-specific commands in the existing modules that were Windows-specific taken out of `Microsoft.PowerShell.Diagnostics` and `Microsoft.PowerShell.Management`. The Windows PowerShell versions of those modules are available in box and marked as Core-compatible too.

All of those probably fall under the `the 5.1 to 6.0 release` classification I imagine

Reply:
Current plan is to only ship with one version of PSGet.